aws/resource_aws_pinpoint_app.go

```diff
@@ -0,0 +1,435 @@
+	}
+
+	d.SetId(*output.ApplicationResponse.Id)
+
+	return resourceAwsPinpointAppUpdate(d, meta)
+}
+
+func resourceAwsPinpointAppUpdate(d *schema.ResourceData, meta interface{}) error {
+	conn := meta.(*AWSClient).pinpointconn
+
+	appSettings := &pinpoint.WriteApplicationSettingsRequest{}
+
+	//if d.HasChange("cloudwatch_metrics_enabled") {
+	//	appSettings.CloudWatchMetricsEnabled = aws.Bool(d.Get("cloudwatch_metrics_enabled").(bool));
+	//}
+
+	if d.HasChange("campaign_hook") {
+		appSettings.CampaignHook = expandCampaignHook(d)
+	}
+
+	if d.HasChange("limits") {
+		appSettings.Limits = expandCampaignLimits(d.Get("limits").(*schema.Set))
+	}
+
+	if d.HasChange("quiet_time") {
+		appSettings.QuietTime = expandQuietTime(d.Get("quiet_time").(*schema.Set))
+	}
+
+	req := pinpoint.UpdateApplicationSettingsInput{
+		ApplicationId:                   aws.String(d.Id()),
+		WriteApplicationSettingsRequest: appSettings,
+	}
+
+	_, err := conn.UpdateApplicationSettings(&req)
+	if err != nil {
+		return err
+	}
+
+	return resourceAwsPinpointAppRead(d, meta)
+}
+
+func resourceAwsPinpointAppRead(d *schema.ResourceData, meta interface{}) error {
+	conn := meta.(*AWSClient).pinpointconn
+
+	log.Printf("[INFO] Reading Pinpoint App Attributes for %s", d.Id())
+
+	app, err := conn.GetApp(&pinpoint.GetAppInput{
+		ApplicationId: aws.String(d.Id()),
+	})
+	if err != nil {
+		if isAWSErr(err, pinpoint.ErrCodeNotFoundException, "") {
+			log.Printf("[WARN] Pinpoint App (%s) not found, error code (404)", d.Id())
+			d.SetId("")
+			return nil
+		}
+
+		return err
+	}
+
+	settings, err := conn.GetApplicationSettings(&pinpoint.GetApplicationSettingsInput{
+		ApplicationId: aws.String(d.Id()),
+	})
+	if err != nil {
+		if isAWSErr(err, pinpoint.ErrCodeNotFoundException, "") {
+			log.Printf("[WARN] Pinpoint App (%s) not found, error code (404)", d.Id())
+			d.SetId("")
+			return nil
+		}
+
+		return err
+	}
+
+	d.Set("name", app.ApplicationResponse.Name)
+	d.Set("application_id", app.ApplicationResponse.Id)
+
+	d.Set("campaign_hook", flattenCampaignHook(settings.ApplicationSettingsResource.CampaignHook))
+	d.Set("limits", flattenCampaignLimits(settings.ApplicationSettingsResource.Limits))
+	d.Set("quiet_time", flattenQuietTime(settings.ApplicationSettingsResource.QuietTime))
+
+	return nil
+}
+
+func resourceAwsPinpointAppDelete(d *schema.ResourceData, meta interface{}) error {
+	conn := meta.(*AWSClient).pinpointconn
+
+	log.Printf("[DEBUG] Pinpoint Delete App: %s", d.Id())
+	_, err := conn.DeleteApp(&pinpoint.DeleteAppInput{
+		ApplicationId: aws.String(d.Id()),
+	})
+	if err != nil {
+		return err
+	}
+	return nil
+}
+
+func expandCampaignHook(d *schema.ResourceData) *pinpoint.CampaignHook {
+	configs := d.Get("campaign_hook").(*schema.Set).List()
+	if configs == nil || len(configs) == 0 {
+		return nil
+	}
+
+	m := configs[0].(map[string]interface{})
+
+	ch := &pinpoint.CampaignHook{}
+
+	if m["lambda_function_name"] != nil {
```

Comment:
If you would to simplify these a little you can use this convention:

```go
if v, ok := m["lambda_function_name"]; ok {
  ch.LambdaFunctionName = aws.String(v.(string))
}
```

Reply:
Fixed in https://github.com/terraform-providers/terraform-provider-aws/pull/5956/commits/d66ab63a1fa5f3ac8ae3598bc8d9717dd9f6a908